### PR TITLE
[FLORA-223] Progressive index import

### PR DESCRIPTION
**Pull request by RaoulHC**
## Proposed changes

This adds an option to import directly from index tarballs rather than needing to unpack them, and then uses the timestamp to avoid re importing anything that is older and should have already been importing. It might be that in the future we want to use the timestamp to grab only that latest part of the index tarball, and avoid checking each cabal file for its timestamp, but we'll still want to be storing the timestamp.

Needs some tests and I'm unsure of the metadata table, but could do with some review in its current form.

I also have a commit to update the freeze file which was out of sync with horizon, but this may be superseded by #386.

## Contributor checklist

- [x] My PR is related to \<insert ticket number> 
- [x] I have read and understood the [CONTRIBUTING guide](https://github.com/flora-pm/flora-server/blob/development/CONTRIBUTING.md)
- [x] I have inserted my change and a link to this PR in the [CHANGELOG](https://github.com/flora-pm/flora-server/blob/development/CHANGELOG.md)

### CODE REVIEW

**Comment by tchoutri** [comment]
@mergifyIO rebase

File: migrations/20230313212225_metadata.sql [hunk 1]
```diff
@@ -0,0 +1,6 @@
+create table if not exists metadata (
```
Comment (tchoutri) [inline migrations/20230313212225_metadata.sql:1]: Let's have instead a table `package_indexes` with the following:

* package_index uuid primary key;
* repository text not null;
* timestamp timestamptz not null;

And an index on the repository field.

:-)

**Review by tchoutri — CHANGES_REQUESTED** [review]

File: migrations/20230313212225_metadata.sql (hunk 1 above)
Comment (RaoulHC) [reply migrations/20230313212225_metadata.sql:1]: Does having a repository make sense quite yet? At some point we'll want to point to a repo and then will make sense, but right now we're importing from a directory or an index file, which doesn't have a repository attached to it. Maybe it makes the most sense to default to `hackage.haskell.org` and perhaps have an argument to the mirror-cli for specifying the repository?

Comment (tchoutri) [reply migrations/20230313212225_metadata.sql]: Good question, at the moment the only repo is Hackage so you can default to it (in the Haskell code, not the SQL). 
A later workflow for CHaP will have to make the distinction between the two, but we're in the clear for now :)

**Comment by tchoutri** [comment]
@RaoulHC do you want to mark this PR as ready for review? Or are there other stuff you want done before the merge?

**Comment by RaoulHC** [comment]
> @RaoulHC do you want to mark this PR as ready for review? Or are there other stuff you want done before the merge?

It could do with some stuff in the test suite, and maybe it wants a flag when importing for whether to update the timestamp table or not, but I think the main logic is good for review. Will try and finish off those bits this week.

File: migrations/20230623168754_add_repo_to_release.sql [hunk 2]
```diff
@@ -0,0 +1,2 @@
+alter table releases
+  add repository text default null;
```
Comment (tchoutri) [inline migrations/20230623168754_add_repo_to_release.sql:2]: I would have thought a foreign key to the package_index_id would be the most appropriate

File: nix/shell-config.nix [hunk 3]
```diff
@@ -7,6 +7,8 @@
     pkgs.ghcid
     pkgs.postgresql_14
     pkgs.souffle
+    pkgs.yarn
+    pkgs.esbuild
```
Comment (tchoutri) [inline nix/shell-config.nix:11]: Nice :+1:

**Review by tchoutri — CHANGES_REQUESTED** [review]
Ok overall I'm very satisfied with this PR. Only one point to resolve. :)

File: migrations/20230623168754_add_repo_to_release.sql (hunk 2 above)
Comment (RaoulHC) [reply migrations/20230623168754_add_repo_to_release.sql:2]: Updated. This does mean we have to add the package index without a timestamp before we start importing releases else you get foreign key violations. We could take the approach of adding/updating the timestamp as we import each release but I think this is probably better.

Comment (tchoutri) [reply migrations/20230623168754_add_repo_to_release.sql:2]: Indeed it's not stupid to have package indexes that we could know about before having importing them, especially if we want to provision the database with public stuff like hackage & chap (and others in the future).

**Review by tchoutri — APPROVED** [review]
LGTM. Thank you very much, it's highly appreciated and extremely valuable.

**Pull request merged**